effects: look for `:terminate_globally` override in backedge termination check

aviatesk opened the pull request:
Now we can "fix" #41694:
```julia
Base.@assume_effects :terminates_globally function issue41694(x)
    res = 1
    1 < x < 20 || throw("bad")
    while x > 1
        res *= x
        x -= 1
    end
    return res
end
@test fully_eliminated() do
    issue41694(2)
end
```

Keno commented on base/expr.jl:
```diff
@@ -532,11 +532,12 @@ macro assume_effects(args...)
             nothrow = true
         elseif setting === :terminates_globally
-            terminates_globally = true
+            terminates_globally = terminates_locally = true
```
Wondering whether it would be better for debugging not to set this override and instead just also check for the global override also on the backedge.

aviatesk committed:
inference: look for `:terminate_globally` override in backedge termination check

Now we can "fix" #41694:
```julia
Base.@assume_effects :terminates_globally function issue41694(x)
    res = 1
    1 < x < 20 || throw("bad")
    while x > 1
        res *= x
        x -= 1
    end
    return res
end
@test fully_eliminated() do
    issue41694(2)
end
```
